### PR TITLE
added testing for emission module

### LINKED ISSUE #1926: [Emissions] Add unit testing to the `Emissions` module (1/2)
There is currently no unit testing of the routines in the `Emissions` module, this should be changed.

## Status

When the routines in the `Emissions` module were written, they needed to be delivered quickly. This resulted in unit tests not being written, and have not been written since the code was added several months ago.

## Intention

Verify correct behavior of the `Emissions` module by writing unit tests for the methods in it.

## Request

Write unit tests for the `Emissions` class located in `emissions.py`.

### Additional Information

There are a lot of methods in `Emissions`, some of them quite large and complex. So unit tests could be split into different PR's to keep the sizes of those PRs more reasonable. Also, if bugs or bad code if found while testing, they should be fixed.

### PR DESCRIPTION
**Pull request by matthew7838**
Add unit tests to `emission.py`, this will be the first part of the PR.

### Context

Issue(s) closed by this pull request: closes #1926 
      
How big are the changes in the PR? Would you nominate it for a major version upgrade?
- [ ] Yes
- [x] No

### What

Adding tests.

### Why

There's no test at all for `emission.py`.

### How

Adding tests for the first half.

### Test plan

This PR is the test. Run with the completed new test suite.

### CODE REVIEW

File: RUFAS/routines/EEE/emissions.py [hunk 1]
```diff
@@ -249,6 +248,7 @@ def _calculate_actual_purchased_feed_emissions(
         actual_land_use_change_emissions = {}
         for feed_id, amount_fed in actual_purchased_feeds.items():
             try:
+                print(feed_id)
```
Comment (ew3361zh) [inline RUFAS/routines/EEE/emissions.py]: remove print statement

File: RUFAS/routines/EEE/emissions.py [hunk 2]
```diff
@@ -442,6 +442,7 @@ def _calculate_emissions_by_field(
         """
         Partitions emissions from the field where crops/feeds were grown to those crops.
         """
+        # skip
```
Comment (ew3361zh) [inline RUFAS/routines/EEE/emissions.py]: What's this comment about?

File: tests/test_EEE_module/test_emissions.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+
+    em._calculate_purchased_feed_emissions(homegrown_feeds)
+
+    assert mock_add.call_count == 3
+    mock_gather_feeds.assert_called_once()
+    mock_calc_actual.assert_called_once_with(homegrown_feeds, purchased_feeds)
+    mock_calc_actual_emission.assert_called_once_with(actual_purchased_feeds)
+
+
+def test_gather_homegrown_feeds_and_fertilizer_apps(mocker: MockerFixture) -> None:
```
Comment (ew3361zh) [inline tests/test_EEE_module/test_emissions.py:146]: add docstring description

File: tests/test_EEE_module/test_emissions.py [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+    mock_add_error = mocker.patch.object(em.om, "add_error")
+    expected = [{"one": 1, "two": 4}, {"one": 2, "two": 5}, {"one": 3, "two": 6}]
+    data = {"one": {"values": [1, 2, 3]}, "two": {"values": [4, 5, 6, 9]}}
+    observed = em._transform_outputs_to_list_of_dicts(data)
+    assert observed == expected
+    mock_add_error.assert_called_once_with(
+        "Found unequal lengths of data while processing simulation outputs for emissions estimation.",
+        "Ignoring extraneous data.",
+        {"class": "EmissionsEstimator", "function": "_transform_outputs_to_list_of_dicts"},
+    )
```
Comment (ew3361zh) [inline tests/test_EEE_module/test_emissions.py]: What do you think about combining these two using parametrization and adding some additional test cases for things like an empty dictionary, empty list values?

**Review by ew3361zh — APPROVED** [review]
Nice work, Matthew. I added a few suggestions but overall this looks great and is a really nice start to getting emissions.py fully covered. I also think it was a good idea to split this into a couple PRs to not overwhelm the size of this PR.

File: tests/test_EEE_module/test_emissions.py (hunk 4 above)
Comment (matthew7838) [reply tests/test_EEE_module/test_emissions.py]: I think it is still better to have different cases for normal and error paths. However, I agree with adding more cases on the error path.

File: RUFAS/routines/EEE/emissions.py [hunk 5]
```diff
@@ -163,15 +163,13 @@ def _gather_ration_feed_totals(self) -> dict[str, float]:
 
         return processed_feeds
 
-    def _transform_outputs_to_list_of_dicts(
-        self, data: dict[str, OutputManager.pool_element_type]
```
Comment (Isgotkowitz) [inline RUFAS/routines/EEE/emissions.py:167]: Why `OutputManager.pool_element_type` -> `Any`?

File: tests/test_EEE_module/test_emissions.py [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+        "Manure Applications": [{"dry_yield": 4, "field_size": 2, "total_dry_yield": 8}],
+        "Manure Requests": [{"dry_yield": 4, "field_size": 2, "total_dry_yield": 8}],
+    }
+    observed = em._gather_homegrown_feeds_and_fertilizer_apps()
+
+    assert observed == expected
+    mock_convert_time.assert_called_once_with(2014, 20)
+    mock_filter_variable.assert_called_once_with(time_filter)
+    assert mock_filter_results.call_count == 4
+    mock_filter_results.assert_has_calls(
```
Comment (Isgotkowitz) [inline tests/test_EEE_module/test_emissions.py]:
```suggestion
    mock_filter_results.assert_has_calls(
```
The call count is checked implicitly when `assert_has_calls` is called.

File: tests/test_EEE_module/test_emissions.py [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+
+def test_gather_ration_feed_totals(mocker: MockerFixture) -> None:
+    """
+    Test that the totals of feeds from rations given to animals in the last 365 days of the simulation are correct.
+    """
+    em = EmissionsEstimator()
+    mock_filter = mocker.patch.object(em.om, "filter_variables_pool", return_value={"test": {"values": [2, 3, 4]}})
+    expected = {"test": 9.0}
+    observed = em._gather_ration_feed_totals()
+    assert observed == expected
```
Comment (Isgotkowitz) [inline tests/test_EEE_module/test_emissions.py]:
```suggestion
    expected = {"test": 9.0}
    
    observed = em._gather_ration_feed_totals()
    
    assert observed == expected
```
It helps readability a lot to separate the test into Arrange-Act-Assert sections

File: tests/test_EEE_module/test_emissions.py [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+        "slice_end": -364,
+    }
+
+    expected = {
+        "Homegrown Feeds": [{"dry_yield": 4, "field_size": 2, "total_dry_yield": 8}],
+        "Fertilizer Applications": [{"dry_yield": 4, "field_size": 2, "total_dry_yield": 8}],
+        "Manure Applications": [{"dry_yield": 4, "field_size": 2, "total_dry_yield": 8}],
+        "Manure Requests": [{"dry_yield": 4, "field_size": 2, "total_dry_yield": 8}],
+    }
+    observed = em._gather_homegrown_feeds_and_fertilizer_apps()
```
Comment (Isgotkowitz) [inline tests/test_EEE_module/test_emissions.py]:
```suggestion
    }
    expected = {
        "Homegrown Feeds": [{"dry_yield": 4, "field_size": 2, "total_dry_yield": 8}],
        "Fertilizer Applications": [{"dry_yield": 4, "field_size": 2, "total_dry_yield": 8}],
        "Manure Applications": [{"dry_yield": 4, "field_size": 2, "total_dry_yield": 8}],
        "Manure Requests": [{"dry_yield": 4, "field_size": 2, "total_dry_yield": 8}],
    }
    
    observed = em._gather_homegrown_feeds_and_fertilizer_apps()
```

File: tests/test_EEE_module/test_emissions.py [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+    )
+    mock_convert_time = mocker.patch.object(Time, "convert_year_jday_to_date", return_value=datetime(2014, 1, 20))
+    mock_filter_results = mocker.patch.object(em, "_filter_results", return_value=[{"dry_yield": 4, "field_size": 2}])
+    time_filter = {
+        "name": "Time Filter",
+        "description": "Collects the date a year before the simulation ended, to be used as a cutoff for deciding "
+        "which crop yields and nutrient applications to estimate emissions for.",
+        "filters": ["Time.(day|calendar_year)"],
+        "slice_start": -365,
+        "slice_end": -364,
```
Comment (Isgotkowitz) [inline tests/test_EEE_module/test_emissions.py:162]: Just a thought, if you declared the filters in `emissions.py` as module constants then you could use them in the tests without having to duplicate them.

File: tests/test_EEE_module/test_emissions.py [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+    )
+
+
+def test_transform_outputs_to_list_of_dicts() -> None:
+    """Test that the function transform data to correct list of dicts."""
+    em = EmissionsEstimator()
+    expected = [{"one": 1, "two": 4}, {"one": 2, "two": 5}, {"one": 3, "two": 6}]
+    data = {"one": {"values": [1, 2, 3]}, "two": {"values": [4, 5, 6]}}
+    observed = em._transform_outputs_to_list_of_dicts(data)
+    assert observed == expected
```
Comment (Isgotkowitz) [inline tests/test_EEE_module/test_emissions.py]:
```suggestion
    expected = [{"one": 1, "two": 4}, {"one": 2, "two": 5}, {"one": 3, "two": 6}]
    data = {"one": {"values": [1, 2, 3]}, "two": {"values": [4, 5, 6]}}
    
    observed = em._transform_outputs_to_list_of_dicts(data)
    
    assert observed == expected
```
I also suggest making "data" and "expected" parameters and testing more cases. What if the data passed is empty? What if it has info maps in it? Possibly combining this test with the one below it and making "expect_error" a parameter as well.

File: tests/test_EEE_module/test_emissions.py [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+
+    mock_totals.assert_called_once_with(homegrown_feeds)
+
+
+def test_calculate_total_homegrown_feed_amounts_by_crop_type(mocker: MockerFixture) -> None:
+    """Tests that the amount of homegrown feeds for all the crop types were calculated correctly."""
+    em = EmissionsEstimator()
+    homegrown_feeds = [
+        {"crop": CropSpecies.CORN_SILAGE, "total_dry_yield": 1200, "dry_matter_content": 0.35},
+        {"crop": CropSpecies.ALFALFA_HAY, "total_dry_yield": 800, "dry_matter_content": 0.9},
```
Comment (Isgotkowitz) [inline tests/test_EEE_module/test_emissions.py:344]: One of `_calculate_total_homegrown_feed_amounts_by_crop_type`'s functionality is to sum the yields of the same crop types. This test does account for that because there is only one corn silage and one alfalfa hay. Could you please add some to the homegrown feeds so that the summation capabilities of this method are tested as well.

**Review by Isgotkowitz — COMMENTED** [review]
This test suite is coming along nicely Matthew, but I think there are some parts of it that could be improved.

**Review by allisterakun — APPROVED** [review]
This PR looks good; great job, Matthew!
Given that it is a two-part process, please clearly indicate which functions are tested and which are left.

File: tests/test_EEE_module/test_emissions.py (hunk 9 above)
Comment (matthew7838) [reply tests/test_EEE_module/test_emissions.py:162]: Great idea, I'll keep that in mind for other parts of the PR. Not much redundancy here so I'll keep it as it is for now.

File: tests/test_EEE_module/test_emissions.py (hunk 10 above)
Comment (matthew7838) [reply tests/test_EEE_module/test_emissions.py]: I do not understand the info maps part. Can you explain a bit more?

Comment (Isgotkowitz) [reply tests/test_EEE_module/test_emissions.py]: When data is pulled out of the of the Output Manager, it may or may not have info maps in it. I.e. you might see 
`{"one": {"values": [1, 2, 3]}, "two": {"values": [4, 5, 6]}}` or `{"one": {"values": [1, 2, 3], "info_maps": [{...}, {...}, {...}]}, "two": {"values": [4, 5, 6], "info_maps": [{...}, {...}, {...}]}}`
